### DISCUSSION TITLE
Is there a memory leak in umqtt.simple or am I doing this wrong?

### DISCUSSION THREAD

**Discussion by Redsandro** [micropython/micropython#13602] [Libraries & Drivers] [answered]
I'm using a Pimoroni Inky Frame 4.0 with a Raspberry Pi Pico W running the [pimoroni-pico](https://github.com/pimoroni/pimoroni-pico#pimoroni-pico-libraries-and-examples-) build of [MicroPython 1.21.0](https://github.com/pimoroni/pimoroni-pico/releases/tag/v1.21.0) to show the latest cat detected by [Frigate](https://github.com/blakeblackshear/frigate#frigate---nvr-with-realtime-object-detection-for-ip-cameras) over MQTT using [`umqtt.simple`](https://github.com/micropython/micropython-lib/tree/7cdf70881519c73667efbc4a61a04d9c1a49babb/micropython/umqtt.simple) 1.3.4 downloaded using Thonny.

So I was subscribed to topic `frigate/pi4/mycamera/#` and only check retained messages on connect one by one (about 14 of them including 4 pictures) until I came across a cat picture at `frigate/pi4/mycamera/cat/snapshot`. Every other picture I'd just ignore, not do anything, assuming no memory references to those messages would exist by the time I do a `gc.collect()` at some later time. 

```python
def on_message_cb(topic, msg):
    if topic == b"frigate/pi4/mycamera/cat/snapshot":
        # Memory consuming code here
    # Ignore every other topic, not making references to `msg`
```

I had however great difficulty showing the smallest picture due to lack of memory with the final `jpegdec.JPEG(graphics).decode()` command throwing `MemoryError: memory allocation failed, allocating 16112 bytes`, even though the `gc.collect()` right before it gave me 40k to play with (although I think those numbers are misleading).

**Now here is the part I don't understand:** 

When I *subscribe* to `frigate/pi4/mycamera/cat/snapshot` directly so there's only 1 callback to handle in stead of a whole bunch of them, the script works flawlessly. I can even double the size of the pictures, and I don't run into `MemoryError`.

Why does this happen? It should not matter. The messages in the ignored callbacks should simply be reclaimed by `gc.collect()`. Am I wrong?

To me it feels like a memory leak, but I know in reality it is infinitely more probable that I made mistakes in my code. I hope someone can point that out. Here's the whole script I called `mqtt_img.py`. For those unfamiliar with the Inky Frame; the "app" (script) is loaded by the launcher `main.py` and `update()` is executed:

```python
import gc, io, os, time, machine, jpegdec
import inky_helper as ih
from umqtt.simple import MQTTClient
from json import dumps, loads
from binascii import hexlify, crc32
from helpers import gc_collect, file_exists
from secrets import secret

# just gc.collect() decorated with a print statement and gc.mem_free()
gc_collect('begin')

# umqtt topics need to be binary for successful comparison
TOPIC_SUB = b""
TOPIC_SUB_ETC = b""
CACHE_FILE='mqtt_img.json'
mqtt_client = None

# Pimoroni Inky Frame specific code
UPDATE_INTERVAL = 5
graphics = None
WIDTH = None
HEIGHT = None
leds = ih.LEDController()

# Called by Inky Frame launcher
def update():
    client = connectMQTT()    
    loops = n = 24 # Checking two dozen messages should be enough
    
    while n > 0:
        print(f'Checking message ({loops-n+1})')
        try:
            client.check_msg()
        except OSError as e:
            if e.args[0] == 9:
                # Done on purpose in the callback to conserve memory
                print("MQTT Client was disconnected.")
                break
        n -= 1
        
    if n == 0:
        print(f'Parsed {loops} messages and nothing happened.')
    else:
        print('Client disconnected. Time to go home.')

def connectMQTT():
    global mqtt_client
    client_id = hexlify(machine.unique_id())
    client_id = f"inky_frame-{client_id.decode('ascii')[:6]}"
    client = MQTTClient(client_id=client_id,
        server=b"mymqtt.localhost",
        port=1883,
        user=b"myuser",
        password=b"mypassword",
        keepalive=7200,
    )

    client.set_callback(on_message_cb)
    client.connect()
    client.subscribe(TOPIC_SUB)
    
    print(f'Connected to MQTT broker, subscribed to {TOPIC_SUB}')
    gc_collect('subscribed')
    mqtt_client = client
    
    return client

def on_message_cb(topic, msg):
    global mqtt_client
    
    if topic == TOPIC_SUB:
        mqtt_client.disconnect()
        buffer = io.BytesIO(msg)
        
        # Calculate hash to compare with cached image
        img_hash = crc32(msg)
        gc_collect('hashed')
        
        # Get old hash from cache
        if file_exists(CACHE_FILE):
            try:
                data = loads(open(CACHE_FILE, "r").read())
                if type(data) is dict:
                    # If the e-ink screen was already on this page, don't refresh old pictures
                    if data['imgHash'] == img_hash and ih.state['run'] == ih.state['prev']:
                        # No update needed; Go back to sleep
                        print('Cache hit. Nothing to do.')
                        return
                else:
                    # File corrupted; Remove
                    os.remove('mqtt_img.json')
            except OSError:
                print('Unexpected OSError.')
                pass
            print('Cache compared, needs redraw')
        # Write cache
        with open(CACHE_FILE, "w") as f:
            f.write(dumps({"imgHash": img_hash}))
            f.flush()
        
        # Rewind buffer and write to file
        buffer.seek(0)
        with open('mqtt_img.jpg', "wb") as f:
            while True:
                chunk = buffer.read(1024)
                if not chunk:
                    break
                f.write(chunk)
                
        # We have a serious need for memory in order to draw the jpeg.
        # Since we call draw(), these vars are still referenced.
        del buffer, msg, img_hash, mqtt_client, data, chunk
        gc_collect('img_draw')
        draw(True)
        
def draw(got_img = False):
    # Override call from main.py 
    if not got_img:
        return
    
    # Set LED status normally done by main.py
    leds.on("warning")
    gc_collect('jpeg')
    jpeg = jpegdec.JPEG(graphics) ### MemoryError here ###
    graphics.set_pen(1)
    graphics.clear()
    jpeg.open_file('mqtt_img.jpg')
    gc_collect('jpdecode')
    jpeg.decode()
    graphics.update()
```

**To summarize:** The script works as long as I subscribe directly to the snapshot topic containing one single snapshot. I run into memory errors if I subscribe to a parent topic that contains multiple snapshots, even if I ignore everything but the one I am looking for.

**Comment by glenn20** [comment, marked as answer]
My first thought is that the other incoming messages would cause increased memory fragmentation, even though you ignore them. Just how fragmented the memory becomes would depend on the rate and pattern of message sizes you receive. If the memory becomes sufficiently fragmented, you will be unable to allocate larger blocks of memory.

PS. Apologies in advance for some other unsolicited advice:
- though not relevant to this particular problem, I think you should be using `client.wait_msg()`, not `client.check_msg()` in `update()` - if I am reading your logic correctly. 
- explicitly close the `CACHE_FILE` in `on_message_cb()` when you open it for reading
- save memory (and memory fragmentation) by just writing the `msg` to the disk file in `update()`, rather than wrapping it in a `BytesIO` instance and writing it in chunks:
    ```py
    with open('mqtt_img.jpg', "wb") as f:
        f.write(msg)
    ```
- and in what I expect is the common case, you may save more memory by drawing the jpeg fom the in-memory `msg`, rather than reading it back from the file (though that depends on just how the app is used).

**Reply by glenn20** [reply]
Also, its not clear from this snippet of code that you need to save the jpeg to a file at all - just render the jpeg direcly from the `msg` buffer (eg. using `jpeg.open_RAM(msg)` - though I havent tested that myself). This will also save wear on your flash.

**Reply by Redsandro** [reply]
> My first thought is that the other incoming messages would cause increased memory fragmentation, even though you ignore them. (...) If the memory becomes sufficiently fragmented, you will be unable to allocate larger blocks of memory.

This hadn't occurred to me. I didn't know that was a thing. Interesting. This is the first time I'm working with a micro controller small enough to require MicroPython, and the first time ever I see a MemoryError. Are there any commands to visualize this, or any methods of defragmenting the memory? E.g. `gc.collect(); gc.defragment()` Or perhaps a way to allocate a sparse block for umqtt's incoming messages so they don't cause further fragmentation?

---

>  Apologies in advance for some other unsolicited advice

Please do!

> - I think you should be using `client.wait_msg()`, not `client.check_msg()` 
 
Yes, I have [considered the difference](https://mpython.readthedocs.io/en/v2.2.1/library/mPython/umqtt.simple.html#umqtt.simple.MQTTClient.wait_msg) and concluded the use cases are waiting for *new* messages versus checking *retained* messages. In this case, I only want to see the retained messages to fetch the last 'snapshot' and let the device go back to deep sleep for a few minutes. The code limiting the amount of messages read was only a guess attempt to prevent further memory errors.

> - explicitly close the `CACHE_FILE` in `on_message_cb()` when you open it for reading

Good catch! `with as` made me lazy and sometimes I forget.

> - save memory (and memory fragmentation) by just writing the `msg` to the disk file in `update()`, rather than wrapping it in a `BytesIO` instance and writing it in chunks

This is my cargo cult showing. I saw this code in the Inky Frame example [nasa_apod.py#L63-L68](https://github.com/pimoroni/pimoroni-pico/blob/6b23c1526df26a5846d4c1f34fdfc8f14c2a1c90/micropython/examples/inky_frame/inkylauncher/nasa_apod.py#L63-L68) and assumed writing in chunks was necessary to prevent memory overhead when writing a file.

```python
socket = urequest.urlopen(IMG_URL)
data = bytearray(1024)
with open(FILENAME, "wb") as f:
	while True:
		if socket.readinto(data) == 0:
			break
		f.write(data)
socket.close()
```
Thanks for pointing out it isn't necessary when the msg is already in memory.

> - just render the jpeg direcly from the `msg` buffer (eg. using `jpeg.open_RAM(msg)`)

To be honest, that was my initial thought as well but I figured if the example `nasa_apod.py` writes the picture to storage directly [before drawing](https://github.com/pimoroni/pimoroni-pico/blob/6b23c1526df26a5846d4c1f34fdfc8f14c2a1c90/micropython/examples/inky_frame/inkylauncher/nasa_apod.py#L84-L92) it to the display, there is probably a good reason for it. I assumed this way it reads and decodes chunks of the file before writing it to the e-ink buffer. Otherwise the decoder, I assumed, would need roughly 844 kB (600 x 480 x 3 / 1024) of memory headroom when there is only about 40 kB available. I will try out your suggestion and I will probably learn that I got it all wrong (again). 😂

**Reply by glenn20** [reply]
> ... the first time ever I see a MemoryError. Are there any commands to visualize this, or any methods of defragmenting the memory? E.g. `gc.collect(); gc.defragment()` Or perhaps a way to allocate a sparse block for umqtt's incoming messages so they don't cause further fragmentation?

Defragmenting memory allocations is extremely complex as it involves moving objects around in memory and updating any references to those objects elsewhere in memory. The micropython memory allocator and gc does not support defragmentation (and neither does Cpython). So, it is best to design your code to reduce defragmentation in the first place. Some of the other tips will substantially help with that. There is also general advice in the micropython docs (eg. https://docs.micropython.org/en/latest/reference/speed_python.html#ram-allocation).

@Josverl provides excellent advice below on how to visualise the memory allocations.

> > * I think you should be using `client.wait_msg()`, not `client.check_msg()`
> 
> Yes, I have [considered the difference](https://mpython.readthedocs.io/en/v2.2.1/library/mPython/umqtt.simple.html#umqtt.simple.MQTTClient.wait_msg) and concluded the use cases are waiting for _new_ messages versus checking _retained_ messages....

Ah, I overlooked the reference to retained messages and agree in this case it should not signifcantly matter.

> > * save memory (and memory fragmentation) by just writing the `msg` to the disk file in `update()`, rather than wrapping it in a `BytesIO` instance and writing it in chunks
> 
> This is my cargo cult showing. I saw this code in the Inky Frame example [nasa_apod.py#L63-L68](https://github.com/pimoroni/pimoroni-pico/blob/6b23c1526df26a5846d4c1f34fdfc8f14c2a1c90/micropython/examples/inky_frame/inkylauncher/nasa_apod.py#L63-L68) and assumed writing in chunks was necessary to prevent memory overhead when writing a file.
> 
> ```python
> socket = urequest.urlopen(IMG_URL)
> data = bytearray(1024)
> with open(FILENAME, "wb") as f:
> 	while True:
> 		if socket.readinto(data) == 0:
> 			break
> 		f.write(data)
> socket.close()
> ```

Yes, this is a good approach when reading an unknown amount of data from a socket. In this case they allocate a single data buffer of reasonable size and reuse that to hold data from the socket before writing it to disk. Only one memory block is allocated and it will work fine even for very large amounts of data. In your case, the `read()` returns a **new** block of memory each time, leading to fragmentation. It is also unnecessary, as you already hold the entire file contents in a memory buffer (`msg`) - just write that to disk.

> To be honest, that was my initial thought as well but I figured if the example `nasa_apod.py` writes the picture to storage directly [before drawing](https://github.com/pimoroni/pimoroni-pico/blob/6b23c1526df26a5846d4c1f34fdfc8f14c2a1c90/micropython/examples/inky_frame/inkylauncher/nasa_apod.py#L84-L92) it to the display, there is probably a good reason for it. I assumed this way it reads and decodes chunks of the file before writing it to the e-ink buffer. Otherwise the decoder, I assumed, would need roughly 844 kB (600 x 480 x 3 / 1024) of memory headroom when there is only about 40 kB available. I will try out your suggestion and I will probably learn that I got it all wrong (again). 😂

Yes, I believe the pimoroni code is designed to work with large jpeg files that may not be possible to hold entirely in memory. In your case, you know your images are entirely contained in memory already. I **believe** your app will work really well if you don't save the images to disk at all, just decode the jpeg directly from the `msg`. Caveat.. I havent verified this, and there may be other requirements that mandate saving to disk. Good luck.

**Reply by Redsandro** [reply]
Thank you for the elaborate educational response. With everything implemented, and saving another kilobyte by shrinking the code (including Pimoroni's scaffolding), everything is running smoothly.

Although if I keep the device running long enough, it could or should eventually run into the same problem because slowly but surely it keeps receiving (and discarding) MQTT messages that may end up fragmenting free memory over time.

**Reply by Redsandro** [reply]
Oh yeah and finally I was hoping to convert the umqtt module from `.py` to `.mpy` because apparently that saves memory too. But I haven't figured out how to run `mpy-cross` on the Pico.

**Reply by robert-hh** [reply]
> But I haven't figured out how to run mpy-cross on the Pico.

You have to run mpy-cross on your PC and upload the .mpy files.

**Reply by Redsandro** [reply]
> You have to run mpy-cross on your PC

@robert-hh don't I need one built specifically for the target MCU and its MicroPython firmware? The Pimoroni "Inky Frame" uses a custom build of MicroPython but they haven't documented how to get, use or build `mpy-cross` for the Inky Frame.

**Reply by glenn20** [reply]
`.mpy` files are generally compatible across different `ports` and, in many cases, micropython release versions. You can see the details in the micropython docs at https://docs.micropython.org/en/latest/reference/mpyfiles.html#versioning-and-compatibility-of-mpy-files. That explains how to identify the bytecode version of your micropython, and which versions are compatible with that bytecode. Then, you **should** be able to download a compatible release of micropython and compile `mpy-cross` (https://github.com/micropython/micropython/blob/master/mpy-cross/README.md). I believe there are plans to make a release of `mpy-cross` that can be distributed without having to build it yourself, but that is not the case yet.

**Reply by glenn20** [reply]
> Thank you for the elaborate educational response. With everything implemented, and saving another kilobyte by shrinking the code (including Pimoroni's scaffolding), everything is running smoothly.

Excellent.. glad to hear it.

> Although if I keep the device running long enough, it could or should eventually run into the same problem because slowly but surely it keeps receiving (and discarding) MQTT messages that may end up fragmenting free memory over time.

Yes, that is true. However, if you are not writing and reading files to/from disk, that should signifcantly reduce memory demand. If it is a significant issue, the [`wait_msg()`](https://github.com/micropython/micropython-lib/blob/master/micropython/umqtt.simple/umqtt/simple.py#L174) function **could** be rewritten to use `readinto()` to save each msg into a single buffer that gets reused for each msg. There are some complexities to that, but it could substantially reduce memory fragmentation.  I'd only consider that if it was necessary.

**Reply by glenn20** [reply]
A simpler solution would probably be to just do a `soft_reset()` when you get an `ENOMEM` exception (assuming it does not happen too frequently).

**Comment by Josverl** [comment]
you can get information on the memory map using 
`import micropython; micropython.mem_info(True)`
to visualize that I have used a Jupyter notebook like  in [this example](https://github.com/Josverl/micropython-magic/blob/main/samples/mem_info.ipynb), or [this plot](https://github.com/Josverl/micropython-magic/blob/main/samples/mem_info-plot.ipynb)
Still that does not give you the info of exactly what object/action is taking space ( although it did help me understand more of the allocation pattern if I slowly step though the allocations map)

**Reply by Redsandro** [reply]
Thanks, this is promising. I have to take some time and learn to understand the output.
